### PR TITLE
Add specific list of supported types in autograd

### PR DESCRIPTION
**Pull request by albanD**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #38740 Make sure cpp factories don't create int Tensor that requires_grad
* #38326 Prevent custom Functions from creating non differentiable type that requires grad
* **#38325 Add specific list of supported types in autograd**

Differential Revision: [D21668739](https://our.internmc.facebook.com/intern/diff/D21668739)

### CODE REVIEW

**Comment by albanD** [comment]
@zdevito now that we're adding new dtype that are supported in the autograd engine, adding a proper valid dtype check.
I found two use cases in the jit code that needed to be updated. Is there more I should check?

File: torch/autograd/__init__.py [hunk 1]
```diff
@@ -3,6 +3,8 @@
 differentiation of arbitrary scalar valued functions. It requires minimal
 changes to the existing code - you only need to declare :class:`Tensor` s
 for which gradients should be computed with the ``requires_grad=True`` keyword.
+As of now, we only support autograd for floating point :class:`Tensor` types (
+half, float, double and bfloat) and complex :class:`Tensor` types (cfloat, cdouble).
```
Comment (gchanan) [inline torch/autograd/__init__.py:7]: nit: bfloat16.

File: torch/csrc/autograd/engine.cpp [hunk 2]
```diff
@@ -630,6 +631,9 @@ void validate_outputs(
       ss << metadata.device() << " but got " << grad_device;
       AT_ERROR(format_error(ss.str()));
     }
+    // The gradient type should match the forward Tensor one
```
Comment (gchanan) [inline torch/csrc/autograd/engine.cpp:634]: not sure what the first line is in reference to, it's not checked here.

File: torch/csrc/autograd/engine.cpp [hunk 3]
```diff
@@ -630,6 +631,9 @@ void validate_outputs(
...
+    // We should not build graph for Tensors that are not differentiable
+    TORCH_INTERNAL_ASSERT(isDifferentiableType(grad.scalar_type()));
```
Comment (gchanan) [inline torch/csrc/autograd/engine.cpp:636]: isn't this already (attempted) to be checked above: `    TORCH_CHECK(isFloatingType(grad.scalar_type()) || (input_is_complex == grad_is_complex));
`

File: torch/csrc/autograd/engine.cpp (hunk 2 above)
Comment (albanD) [reply torch/csrc/autograd/engine.cpp:634]: Yes, I will remove this.

File: torch/csrc/autograd/engine.cpp (hunk 3 above)
Comment (albanD) [reply torch/csrc/autograd/engine.cpp:636]: The semantic is a bit weird here because, most notably for quantized training, we allow backrpopagating gradient of a different type from the forward Tensor.
The check you mention above is here to make sure that all the casting operations are valid.
This one is to make sure that the final Tensor we get after all the casting is a valid dtype. (it should always be as it should match the forward dtype which has to be differentiable for us to get here).

Comment (gchanan) [reply torch/csrc/autograd/engine.cpp:635]: is the quantized training thing documented anywhere?

**Review by gchanan — APPROVED** [review]
